src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -143,38 +144,54 @@ void reconnect() throws TimeoutException, ExecutionException, InterruptedExcepti
         // Synchronize here instead of method signature to make mockito work without deadlocking
         synchronized (this) {
             logger.atInfo().log("Reconnecting MQTT client most likely due to device configuration change");
-            disconnect();
+            disconnect().get(getTimeout(), TimeUnit.MILLISECONDS);
             connect().get(getTimeout(), TimeUnit.MILLISECONDS);
         }
     }
 
-    @SuppressWarnings("PMD.NullAssignment")
     protected synchronized CompletableFuture<Boolean> connect() {
         if (connection != null) {
             return CompletableFuture.completedFuture(true);
         }
+        // For the initial connect, client connects with cleanSession=true and disconnects.
+        // This deletes any previous session information maintained by IoT Core.
+        // For subsequent connects, the client connects with cleanSession=false
+        CompletableFuture<Void> voidCompletableFuture =  CompletableFuture.completedFuture(null);
+        if (initalConnect.get()) {
```

Comment:
While the things inside these callbacks are running you need to ensure that no other thread can come in and call connect() and execute these things at the same time. This is possible to do, but difficult to get right. I implemented it in an earlier closed PR by saving one future and returning it to anyone who called connect() so that all callers were waiting on the same connect() future.

Reply:
Changed to ensure there that only one thread can request a connect and all subsequent calls to connect will wait for an ongoing connect operation.